packages/cli-repl/src/logger.ts

```diff
@@ -1,53 +1,189 @@
+        userId,
+        event: 'mongosh:connect',
+        properties: { sessionID, connectionUri, ...args }
+      });
+    }
   });
 
-  bus.on('mongosh:error', function(error) {
+  bus.on('mongosh:new-user', function(id, enableTelemetry) {
+    userId = id;
+    telemetry = enableTelemetry;
+    if (telemetry) analytics.identify({ userId });
+  })
+
+  bus.on('mongosh:update-user', function(id, enableTelemetry) {
+    userId = id;
+    telemetry = enableTelemetry;
+    log.info('mongosh:update-user', { enableTelemetry })
+  })
+
+
+  bus.on('mongosh:error', function(error: ErrorEvent) {
     log.error(error);
+
+    if (telemetry) {
+      analytics.track({
+        userId,
+        event: 'mongosh:error',
+        properties: { error }
+      });
+    }
   });
 
   bus.on('mongosh:help', function() {
     log.info('mongosh:help');
+
+    if (telemetry) {
+      analytics.track({
+        userId,
+        event: 'mongosh:help'
+      });
+    }
   });
 
-  bus.on('mongosh:rewrittenAsyncInput', function(inputInfo) {
-    log.info('mongosh:rewrittenAsyncInput', inputInfo);
+  bus.on('mongosh:rewritten-async-input', function(args: AsyncRewriterEvent) {
+    log.info('mongosh:rewritten-async-input', args);
   });
 
-  bus.on('mongosh:use', function(args) {
-    log.info(args);
+  bus.on('mongosh:use', function(args: UseEvent) {
+    log.info('mongosh:use', args);
+
+    if (telemetry) {
+      analytics.track({
+        userId,
+        event: 'mongosh:use'
+      });
+    }
   });
 
-  bus.on('mongosh:show', function(args) {
-    log.info(args);
+  bus.on('mongosh:show', function(args: ShowEvent) {
+    log.info('mongosh:show', args);
+
+    if (telemetry) {
+      analytics.track({
+        userId,
+        event: 'mongosh:show',
+        properties: { method: args.method }
+      });
+    }
   });
 
-  bus.on('mongosh:it', function(args) {
-    log.info(args);
+  bus.on('mongosh:it', function() {
+    log.info('mongosh:it');
+
+    if (telemetry) {
+      analytics.track({
+        userId,
+        event: 'mongosh:it'
+      });
+    }
   });
 
   bus.on('mongosh:setCtx', function(args) {
-    log.info(args)
+    log.info('mongosh:setCtx')
   })
 
-  bus.on('mongosh:api-call', function(args) {
-    log.info(redactInfo(args));
+  bus.on('mongosh:api-call', function(args: ApiEvent) {
+    log.info('mongosh:api-call', redactInfo(args));
+
+    // analytics properties to include if they are present in an api-call
+    let properties: ApiEvent = {};
+    properties.arguments = {};
+    if (args.method) properties.method = args.method;
+    if (args.class) properties.class = args.class;
+    if (args.arguments) properties.arguments = properties.arguments;
+
+    if (telemetry) {
+      analytics.track({
+        userId,
+        event: 'mongosh:api-call',
+        properties: redactInfo(properties)
+      });
+    }
   });
 }
 
-export default logger;
+// set up a noop, in case we are not able to connect to segment.
+function NoopAnalytics() {}
```

Comment:
Can't `class` be used here?

Reply:
what's the syntax for a noop class? I've only ever written it like this before.

Reply:
I meant this:

```
class NoopAnalytics {
   identify() {}
   track() {}
}
```

I was wondering if you had some form of typescript issue with that. Just a formality anyway.

Reply:
oh no, nothing to do with typescript! I am just old school and didn't know the new and fancy syntax

Reply:
in an unexpected turn of events, typescript does not, in fact like the class syntax.